src/oidcendpoint/authz/__init__.py

```diff
@@ -1,63 +1,101 @@
 import logging
 import sys
+from typing import Optional
+from typing import Union
 
-from oidcendpoint import sanitize
-from oidcendpoint.cookie import cookie_value
+from oidcmsg.message import Message
+
+from oidcendpoint.session.grant import Grant
 
 logger = logging.getLogger(__name__)
 
 
 class AuthzHandling(object):
     """ Class that allow an entity to manage authorization """
 
-    def __init__(self, endpoint_context, **kwargs):
+    def __init__(self, endpoint_context, grant_config=None, **kwargs):
         self.endpoint_context = endpoint_context
         self.cookie_dealer = endpoint_context.cookie_dealer
-        self.permdb = {}
+        self.grant_config = grant_config or {}
         self.kwargs = kwargs
 
-    def __call__(self, *args, **kwargs):
-        return ""
+    def usage_rules(self, client_id):
+        if "usage_rules" in self.grant_config:
+            _usage_rules = copy.deepcopy(self.grant_config["usage_rules"])
+        else:
+            _usage_rules = {}
 
-    def set(self, uid, client_id, permission):
         try:
-            self.permdb[uid][client_id] = permission
+            _per_client = self.endpoint_context.cdb[client_id]["token_usage_rules"]
         except KeyError:
-            self.permdb[uid] = {client_id: permission}
-
-    def permissions(self, cookie=None, **kwargs):
-        if cookie is None:
-            return None
+            pass
         else:
-            logger.debug("kwargs: %s" % sanitize(kwargs))
-
-            val = self.cookie_dealer.get_cookie_value(cookie)
-            if val is None:
-                return None
+            if _usage_rules:
+                for _token_type, _rule in _usage_rules.items():
+                    _pc = _per_client.get(_token_type)
+                    if _pc:
+                        _rule.update(_pc)
+                for _token_type, _rule in _per_client.items():
+                    if _token_type not in _usage_rules:
+                        _usage_rules[_token_type] = _rule
             else:
-                b64, _ts, typ = val
+                _usage_rules = _per_client
 
-            info = cookie_value(b64)
-            return self.get(info["sub"], info["client_id"])
+        return _usage_rules
 
-    def get(self, uid, client_id):
+    def usage_rules_for(self, client_id, token_type):
+        _token_usage = self.usage_rules(client_id=client_id)
         try:
-            return self.permdb[uid][client_id]
+            return _token_usage[token_type]
         except KeyError:
-            return None
+            return {}
 
+    def __call__(self, session_id: str, request: Union[dict, Message],
+                 resources: Optional[list] = None) -> Grant:
+        args = self.grant_config.copy()
 
-class Implicit(AuthzHandling):
-    def __init__(self, endpoint_context, permission="implicit"):
-        AuthzHandling.__init__(self, endpoint_context)
-        self.permission = permission
+        scope = request.get("scope")
```

Comment:
`request.get("scope") ` may return a space separated list, we should split the scopes before passing them to args

Reply:
This is why I've always questioned why scope was this stupid space separated 'list'.
But the powers that be wouldn't listen to me.
Anyway if request is a Message instance you will **always** get a list. If request is a dict then all bets are off.